zebra-consensus/src/primitives/halo2.rs

```diff
@@ -0,0 +1,283 @@
+    }
+}
+
+#[derive(Default)]
+pub struct BatchVerifier {
+    queue: Vec<Item>,
+}
+
+impl BatchVerifier {
+    pub fn queue(&mut self, item: Item) {
+        self.queue.push(item);
+    }
+
+    pub fn verify<R: RngCore + CryptoRng>(
+        self,
+        _rng: R,
+        vk: &VerifyingKey,
+    ) -> Result<(), halo2::plonk::Error> {
+        for item in self.queue {
+            item.verify_single(vk)?;
+        }
+
+        Ok(())
+    }
+}
+
+// === END TEMPORARY BATCH HALO2 SUBSTITUTE ===
+
+impl From<zebra_chain::orchard::ShieldedData> for Item {
+    fn from(shielded_data: zebra_chain::orchard::ShieldedData) -> Item {
+        use orchard::{circuit, note, primitives::redpallas, tree, value};
+
+        let anchor = tree::Anchor::from_bytes(shielded_data.shared_anchor.into()).unwrap();
+
+        let enable_spend = shielded_data
+            .flags
+            .contains(zebra_chain::orchard::Flags::ENABLE_SPENDS);
+        let enable_output = shielded_data
+            .flags
+            .contains(zebra_chain::orchard::Flags::ENABLE_OUTPUTS);
+
+        let instances = shielded_data
+            .actions()
+            .map(|action| {
+                circuit::Instance::from_parts(
+                    anchor,
+                    value::ValueCommitment::from_bytes(&action.cv.into()).unwrap(),
+                    note::Nullifier::from_bytes(&action.nullifier.into()).unwrap(),
+                    redpallas::VerificationKey::<redpallas::SpendAuth>::try_from(<[u8; 32]>::from(
+                        action.rk,
+                    ))
+                    .expect("should be a valid redpallas spendauth verification key"),
+                    note::ExtractedNoteCommitment::from_bytes(&action.cm_x.into()).unwrap(),
+                    enable_spend,
+                    enable_output,
+                )
+            })
+            .collect();
+
+        Item {
+            instances,
+            proof: orchard::circuit::Proof::new(shielded_data.proof.0),
+        }
+    }
+}
+
+/// An error that may occur when verifying [Halo2 proofs of Zcash Orchard Action
+/// descriptions][actions].
+///
+/// [actions]: https://zips.z.cash/protocol/protocol.pdf#actiondesc
+// TODO: if halo2::plonk::Error gets the std::error::Error trait derived on it,
+// remove this and just wrap `halo2::plonk::Error` as an enum variant of
+// `crate::transaction::Error`, which does the trait derivation via `thiserror`
+#[derive(Clone, Debug, Error, Eq, PartialEq)]
+pub enum Halo2Error {
+    /// The constraint system is not satisfied.
+    ConstraintSystemFailure,
+    /// Catchall for now until https://github.com/zcash/halo2/pull/394 is merged
+    Other,
+}
+
+impl fmt::Display for Halo2Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        match self {
+            Halo2Error::ConstraintSystemFailure => write!(f, "Constraint system failure"),
+            _ => write!(f, "Unknown Halo2 error"),
+        }
+    }
+}
```

Comment:
Optional idea: sorry to bring this up again, but I realized we could have the `Display` implementation generated by `thiserror`.

```suggestion
pub enum Halo2Error {
    /// The constraint system is not satisfied.
    #[error("Constraint system failure")]
    ConstraintSystemFailure,
    
    /// Catchall for now until https://github.com/zcash/halo2/pull/394 is merged
    #[error("Unknown Halo2 error")]
    Other,
}
```

Reply:
My suggestions finally got merged in https://github.com/zcash/halo2/pull/338/files#diff-ce29ac47e4ad4e869b74fd280b52b536e4967606470ae4f64261c50c32e4404f, so if all the other dependencies work out, I might be able to use `halo2::plonk::Error` after all - if not, I'll use your suggestions

Reply:
Done in 25082b7cdd08ac5489225ee1f742f01bb549d9f7 for now